File: 백준/Gold/2504. 괄호의 값/괄호의 값.py
s = list(input())

result = []
count = 1
ans = 0

for i in range(len(s)):
    if s[i] == '[':
        result.append(s[i])
        count *= 3

    elif s[i] == '(':
        result.append(s[i])
        count *= 2

    elif s[i] == ']':
        if not result or result[-1] == '(':
            ans = 0
            break
        if s[i - 1] == '[':
            ans += count
        count //= 3
        result.pop()

    elif s[i] == ')':
        if not result or result[-1] == '[':
            ans = 0
            break
        if s[i - 1] == '(':
            ans += count
        count //= 2
        result.pop()

if result:
    ans = 0
print(ans)

File: 백준/Gold/2504. 괄호의 값/괄호의 값.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;


public class Main {
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		char arr[] = br.readLine().toCharArray();
		Stack<Character> st = new Stack<>(); 
		
		int ans = 0; 
		int cost = 1;
		
		for(int i=0; i<arr.length;i++) {
			char c = arr[i];
			if(c=='(' || c=='[') {
				st.push(c);
				cost *= (c=='('? 2 : 3); 
			}
			else {
				if(st.isEmpty()) {
					ans=0; 
					break;
				}
				else {
					if(c==')') {
						if(arr[i-1]=='(')
							ans += cost;
						cost /=2;
						if(st.peek()=='(')
							st.pop();
					}
					else if(c==']') {
						if(arr[i-1]=='[')
							ans+=cost;
						cost/=3;
						if(st.peek()=='[')
							st.pop();
					}
				}
			}
			
		}
		if(!st.isEmpty()) ans=0;
		
		System.out.println(ans);	
		
	}
}
